File: web/src/views/visitors.js
import { h, fill, badge, chipRow, photoSlot } from '../components/ui.js';
import * as S from '../state.js';
import { opponents } from '../data/mock.js';

let activeId = opponents[0].id;

/* The room down the tunnel from the Hearth: bare block walls, a bench an inch
   out of true, a team sheet taped over last week's. The dishevelment is
   cosmetic — tilts stay under 0.6°, nothing animates, and every piece of text
   sits on a solid surface at full contrast. */

const CONCRETE = 'repeating-linear-gradient(180deg, rgba(120,113,108,.07) 0 38px, rgba(120,113,108,.12) 38px 39px)';

export function visitorsView() {
  const wrap = h('div', { class: 'view shell', style: { paddingTop: '16px' } });

  const paint = () => {
    const op = opponents.find((o) => o.id === activeId) ?? opponents[0];

    fill(wrap,
      room(),
      h('div', { style: { marginTop: '14px' } },
        chipRow(opponents.map((o) => ({ value: o.id, label: o.chip })), activeId,
          (v) => { activeId = v; paint(); }, { stone: true }),
      ),
      dossier(op),
    );
  };

  paint();
  return wrap;
}

function room() {
  return h('section', {
    style: {
      borderRadius: 'var(--radius-lg)',
      border: '1px solid var(--hairline-strong)',
      background: `${CONCRETE}, var(--surface-sunk)`,
      padding: '18px 16px',
      position: 'relative',
    },
  },
    // Taped sign, slightly off true.
    h('div', {
      style: {
        display: 'inline-block',
        transform: 'rotate(-0.5deg)',
        background: '#E7E2D4',
        color: '#1C1917',
        border: '1px solid #C9C2AE',
        padding: '6px 14px',
        fontFamily: 'var(--font-mono)',
        fontSize: '11px',
        letterSpacing: '.14em',
        fontWeight: '500',
      },
    }, 'AWAY · 5,280 FT'),

    h('h1', { style: { fontSize: '22px', fontWeight: '800', marginTop: '12px' } },
      "The Visiting Club's Dressing Room"),
    h('p', { style: { margin: '8px 0 0', fontSize: '14px', color: 'var(--ink-soft)', maxWidth: '58ch' } },
      'Down the tunnel from the Hearth: bare block walls, a bench an inch out of true, and a team sheet taped over last week’s. Everything you need on whoever is in town.'),

    // The team sheet: every club, both legs.
    h('div', {
      style: {
        marginTop: '16px',
        transform: 'rotate(0.4deg)',
        background: 'var(--surface)',
        border: '1px solid var(--hairline-strong)',
        padding: '14px',
        borderRadius: '4px',
      },
    },
      h('p', { class: 'eyebrow' }, 'Team sheet'),
      h('div', { style: { marginTop: '8px' } },
        opponents.map((o) => h('div', {
          style: {
            display: 'flex', justifyContent: 'space-between', gap: '10px',
            padding: '7px 0', borderBottom: '1px solid var(--hairline)', fontSize: '13px',
          },
        },
          h('span', { style: { fontWeight: '600' } }, o.club),
          h('span', { style: { fontFamily: 'var(--font-mono)', color: 'var(--ink-mute)', textAlign: 'right' } },
            `HOME ${o.homeDate} · AWAY ${o.awayDate}`),
        )),
      ),
      h('p', { style: { margin: '10px 0 0', fontSize: '12px', color: 'var(--ink-mute)' } },
        'Away dates are for travel planning only — they sit outside the season package.'),
    ),
  );
}

function dossier(op) {
  return h('section', { style: { marginTop: '14px' } },
    h('article', { class: 'card' },
      h('div', { style: { display: 'flex', justifyContent: 'space-between', gap: '10px', flexWrap: 'wrap' } },
        h('div', null,
          h('h2', { style: { fontSize: '19px', fontWeight: '800' } }, op.club),
          h('p', { style: { margin: '4px 0 0', fontFamily: 'var(--font-mono)', fontSize: '12px', color: 'var(--ink-mute)' } },
            `HOME ${op.homeDate} · AWAY ${op.awayDate}`),
          h('p', { style: { margin: '2px 0 0', fontSize: '13px', color: 'var(--ink-mute)' } },
            `Away leg at ${op.awayVenue} — outside the package.`),
        ),
        h('div', null,
          h('p', { class: 'eyebrow' }, 'Form'),
          h('p', { style: { margin: '4px 0 0', fontFamily: 'var(--font-mono)', fontWeight: '500', letterSpacing: '.18em' } }, op.form),
        ),
      ),

      h('div', { style: { display: 'flex', gap: '8px', marginTop: '14px', flexWrap: 'wrap' } },
        op.quickStats.map(([k, v]) => h('div', {
          class: 'card card--sunk',
          style: { padding: '8px 12px', borderRadius: '10px', flex: '1 1 96px' },
        },
          h('p', { class: 'eyebrow' }, k),
          h('p', { style: { margin: '2px 0 0', fontFamily: 'var(--font-display)', fontWeight: '700', fontSize: '18px' } }, v),
        )),
      ),

      h('div', {
        style: {
          marginTop: '14px', padding: '12px 14px', borderRadius: '10px',
          background: 'rgba(200,75,49,.1)', borderLeft: '3px solid var(--summit-sandstone)',
        },
      },
        h('p', { class: 'eyebrow', style: { color: 'var(--summit-sandstone)' } }, 'Halftime read'),
        h('p', { style: { margin: '5px 0 0', fontSize: '14px', color: 'var(--ink-soft)' } }, op.halftime),
      ),

      h('p', { style: { margin: '14px 0 0', fontSize: '14px' } },
        h('strong', null, 'Shape: '), op.shape),
    ),

    h('p', { class: 'eyebrow', style: { margin: '16px 0 8px' } }, 'Dressing room occupants'),
    h('div', { class: 'grid-auto', style: { alignItems: 'start' } }, op.players.map(pegCard)),
  );
}

/** Each player reads as a kit peg you tap to pull the shirt off the hook. */
function pegCard(p) {
  let open = false;
  const card = h('article', {
    class: 'card',
    style: { transform: `rotate(${(p.num % 2 ? -0.35 : 0.3)}deg)` },
  });

  const paint = () => {
    fill(card,
      h('div', { style: { display: 'flex', gap: '12px' } },
        photoSlot(p.id, 84, S.get().photos[p.id], (url) => { S.setPhoto(p.id, url); paint(); }),
        h('div', { style: { flex: '1', minWidth: '0' } },
          h('p', { style: { margin: '0', display: 'flex', alignItems: 'center', gap: '8px', flexWrap: 'wrap' } },
            badge(`#${p.num}`, 'gold'),
            h('a', {
              href: `https://en.wikipedia.org/w/index.php?search=${encodeURIComponent(p.name)}`,
              target: '_blank', rel: 'noopener noreferrer',
              style: { fontFamily: 'var(--font-display)', fontWeight: '700', fontSize: '15px' },
            }, p.name),
            p.danger ? badge('DANGER', 'ember') : null,
          ),
          h('p', { style: { margin: '3px 0 0', fontFamily: 'var(--font-mono)', fontSize: '12px', color: 'var(--ink-mute)' } }, p.pos),
          h('div', { style: { display: 'flex', gap: '10px', marginTop: '6px', flexWrap: 'wrap' } },
            h('a', { href: '#', style: { fontSize: '13px' }, onclick: (e) => e.preventDefault() }, 'Club bio ↗'),
            h('a', {
              href: `https://en.wikipedia.org/w/index.php?search=${encodeURIComponent(p.name)}`,
              target: '_blank', rel: 'noopener noreferrer',
              style: { fontSize: '13px' },
            }, 'Wikipedia ↗'),
          ),
        ),
      ),
      h('button', {
        class: 'btn btn--ghost btn--block',
        type: 'button',
        style: { marginTop: '10px' },
        'aria-expanded': String(open),
        onclick: () => { open = !open; paint(); },
      }, open ? 'Back on the hook' : 'Pull it off the hook'),
      open
        ? h('p', { style: { margin: '10px 0 0', fontSize: '14px', color: 'var(--ink-soft)' } }, p.note)
        : null,
    );
  };

  paint();
  return card;
}

File: web/build_src/views/visitors.py
"""Visitors: the visiting club's dressing room. Every opponent's dossier and
roster is pre-rendered; switching the chip just shows/hides which one is
visible. Tilts stay under 0.6deg and nothing animates, per the design brief.
"""
from __future__ import annotations

import sys, os
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from markup import h
from components import badge, photo_slot, bio_links
from data import OPPONENTS

CONCRETE = "repeating-linear-gradient(180deg, rgba(120,113,108,.07) 0 38px, rgba(120,113,108,.12) 38px 39px)"


def render():
    return h(
        "div", {"cls": "view shell", "style": {"paddingTop": "16px"}, "data-tab": "visitors"},
        room(),
        h("div", {"style": {"marginTop": "14px"}, "cls": "scroll-row", "role": "tablist"},
          [h("button", {"cls": "chip chip--stone", "type": "button", "role": "tab",
                        "aria-selected": "true" if i == 0 else "false",
                        "data-role": "opponent", "data-value": op["id"]}, op["chip"])
           for i, op in enumerate(OPPONENTS)]),
        [h("div", {"data-opponent-block": op["id"], "hidden": i != 0}, dossier(op))
         for i, op in enumerate(OPPONENTS)],
    )


def room():
    rows = [h("div", {"style": {"display": "flex", "justifyContent": "space-between", "gap": "10px",
                                  "padding": "7px 0", "borderBottom": "1px solid var(--hairline)", "fontSize": "13px"}},
              h("span", {"style": {"fontWeight": "600"}}, op["club"]),
              h("span", {"style": {"fontFamily": "var(--font-mono)", "color": "var(--ink-mute)", "textAlign": "right"}},
                f'HOME {op["home_date"]} · AWAY {op["away_date"]}'))
             for op in OPPONENTS]

    return h(
        "section", {"style": {"borderRadius": "var(--radius-lg)", "border": "1px solid var(--hairline-strong)",
                               "background": f"{CONCRETE}, var(--surface-sunk)", "padding": "18px 16px", "position": "relative"}},
        h("div", {"style": {"display": "inline-block", "transform": "rotate(-0.5deg)", "background": "#E7E2D4",
                             "color": "#1C1917", "border": "1px solid #C9C2AE", "padding": "6px 14px",
                             "fontFamily": "var(--font-mono)", "fontSize": "11px", "letterSpacing": ".14em", "fontWeight": "500"}},
          "AWAY · 5,280 FT"),
        h("h1", {"style": {"fontSize": "22px", "fontWeight": "800", "marginTop": "12px"}}, "The Visiting Club's Dressing Room"),
        h("p", {"style": {"margin": "8px 0 0", "fontSize": "14px", "color": "var(--ink-soft)", "maxWidth": "58ch"}},
          "Down the tunnel from Basecamp: bare block walls, a bench an inch out of true, and a team sheet taped over last week's. Everything you need on whoever is in town."),
        h("div", {"style": {"marginTop": "16px", "transform": "rotate(0.4deg)", "background": "var(--surface)",
                             "border": "1px solid var(--hairline-strong)", "padding": "14px", "borderRadius": "4px"}},
          h("p", {"cls": "eyebrow"}, "Team sheet"),
          h("div", {"style": {"marginTop": "8px"}}, rows),
          h("p", {"style": {"margin": "10px 0 0", "fontSize": "12px", "color": "var(--ink-mute)"}},
            "Away dates are for travel planning only — they sit outside the season package.")),
    )


def dossier(op):
    empty_players_note = h(
        "p", {"style": {"margin": "0", "fontSize": "14px", "color": "var(--ink-mute)"}},
        "Full team sheet not published here yet — see the ",
        h("a", {"href": op["match_url"], "target": "_blank", "rel": "noopener noreferrer"},
          "latest meeting"),
        " for who played.",
    )
    return h(
        "section", {"style": {"marginTop": "14px"}},
        h("article", {"cls": "card"},
          h("div", None,
            h("h2", {"style": {"fontSize": "19px", "fontWeight": "800"}}, op["club"]),
            h("p", {"style": {"margin": "4px 0 0", "fontFamily": "var(--font-mono)", "fontSize": "12px", "color": "var(--ink-mute)"}},
              f'HOME {op["home_date"]} · AWAY {op["away_date"]}'),
            h("p", {"style": {"margin": "2px 0 0", "fontSize": "13px", "color": "var(--ink-mute)"}},
              f'Away leg at {op["away_venue"]} — outside the package.')),
          h("div", {"style": {"marginTop": "14px", "padding": "12px 14px", "borderRadius": "10px",
                               "background": "rgba(200,75,49,.1)", "borderLeft": "3px solid var(--summit-sandstone)"}},
            h("p", {"cls": "eyebrow", "style": {"color": "var(--summit-sandstone)"}}, "This season"),
            h("p", {"style": {"margin": "5px 0 0", "fontSize": "14px", "color": "var(--ink-soft)"}}, op["recent_result"]))),
        h("p", {"cls": "eyebrow", "style": {"margin": "16px 0 8px"}}, "Dressing room occupants"),
        h("div", {"cls": "grid-auto", "style": {"alignItems": "start"}}, [peg_card(p, op) for p in op["players"]])
        if op["players"] else empty_players_note,
    )


def peg_card(p, op):
    tilt = -0.35 if p["num"] % 2 else 0.3
    return h(
        "article", {"cls": "card", "style": {"transform": f"rotate({tilt}deg)"}},
        h("div", {"style": {"display": "flex", "gap": "12px"}},
          photo_slot(p["id"], 84),
          h("div", {"style": {"flex": "1", "minWidth": "0"}},
            h("p", {"style": {"margin": "0", "display": "flex", "alignItems": "center", "gap": "8px", "flexWrap": "wrap"}},
              badge(f'#{p["num"]}', "gold"),
              h("a", {"href": f'https://en.wikipedia.org/w/index.php?search={_urlenc(p["name"])}',
                       "target": "_blank", "rel": "noopener noreferrer",
                       "style": {"fontFamily": "var(--font-display)", "fontWeight": "700", "fontSize": "15px"}}, p["name"]),
              badge("DANGER", "ember") if p["danger"] else None),
            h("p", {"style": {"margin": "3px 0 0", "fontFamily": "var(--font-mono)", "fontSize": "12px", "color": "var(--ink-mute)"}}, p["pos"]),
            bio_links(p["name"], op["match_url"], club_label="Latest meeting"))),
        h("button", {"cls": "btn btn--ghost btn--block", "type": "button", "style": {"marginTop": "10px"},
                     "aria-expanded": "false", "data-role": "expand", "data-target": f'peg-detail-{p["id"]}',
                     "data-label-open": "Pull it off the hook", "data-label-close": "Back on the hook"},
          "Pull it off the hook"),
        h("p", {"style": {"margin": "10px 0 0", "fontSize": "14px", "color": "var(--ink-soft)"}, "id": f'peg-detail-{p["id"]}', "hidden": True},
          p["note"]),
    )


def _urlenc(text):
    from urllib.parse import quote
    return quote(text)
